src/raftstore/store/snap.rs

```diff
@@ -618,6 +620,1006 @@ mod v1 {
+                if size != cf_file.size {
+                    return Err(io::Error::new(ErrorKind::Other,
+                                              format!("snapshot file {} for cf {} size \
+                                                       mismatches, real size {}, expected size \
+                                                       {}",
+                                                      cf_file.path,
+                                                      cf_file.cf,
+                                                      size,
+                                                      cf_file.size)));
+                }
+                let checksum = try!(calc_checksum(&cf_file.tmp_path));
+                if checksum != cf_file.checksum {
+                    return Err(io::Error::new(ErrorKind::Other,
+                                              format!("snapshot file {} for cf {} checksum \
+                                                       mismatches, real checksum {}, expected \
+                                                       checksum {}",
+                                                      cf_file.path,
+                                                      cf_file.cf,
+                                                      checksum,
+                                                      cf_file.checksum)));
+                }
+            }
+            let mut total_size = 0;
+            for cf_file in &mut self.cf_files {
+                try!(fs::rename(&cf_file.tmp_path, &cf_file.path));
+                total_size += cf_file.size;
+            }
+            // write meta file
+            let mut v = vec![];
+            try!(self.meta_file.data.write_to_vec(&mut v));
+            if let Err(e) = self.meta_file.file.take().unwrap().write_all(&v[..]) {
+                return Err(io::Error::new(ErrorKind::Other, e.description()));
+            }
+            try!(fs::rename(&self.meta_file.tmp_path, &self.meta_file.path));
+            let mut size_track = self.size_track.wl();
+            *size_track = size_track.saturating_add(total_size);
+            Ok(())
+        }
+
+        fn apply(&mut self, context: &mut ApplyContext) -> Result<()> {
+            for (cf, path, size) in box_try!(self.list_cf_path_sizes()) {
+                if size == 0 {
+                    // Skip ingesting empty sst file since that would cause an error
+                    // "Corruption: file is too short to be an sstable"
+                    continue;
+                }
+                try!(check_abort(&context.abort));
+
+                let cf_handle = box_try!(rocksdb::get_cf_handle(&context.db, &cf));
+                let ingest_opt = IngestExternalFileOptions::new().move_files(true);
+                if let Err(e) = context.db
+                    .ingest_external_file_cf(cf_handle, &ingest_opt, &[&path]) {
+                    return Err(Error::Other(box_err!(e)));
```

Comment:
In this case, the region data in db will be dirty. And the region will retry to apply this snapshot or a newer snapshot until it succeeds in applying a whole snapshot.